src/components/sharing/util.js

```diff
@@ -0,0 +1,431 @@
+                            source_id: user.source_id,
+                            id: user.id,
+                        },
+                    };
+                },
+                fmtSharingFn: (resource) => {
+                    return {
+                        tool_id: resource.id || resource.tool_id,
+                    };
+                },
+                fmtUnsharingFn: (resource) => resource.id || resource.tool_id,
+            };
+        }
+        default:
+            return null;
+    }
+};
+
+// A user can have all or no resources.  Check for each resource type and
+// return a list containing them all
+const getUserResourceList = (user) => {
+    return [
+        ...(user[TYPE.DATA] || []),
+        ...(user[TYPE.ANALYSES] || []),
+        ...(user[TYPE.APPS] || []),
+        ...(user[TYPE.TOOLS] || []),
+    ];
+};
+
+const addDisplayPermissions = (userMap, resourceTotal) => {
+    return Object.keys(userMap).reduce(
+        (acc, userId) => {
+            const user = acc[userId];
+            const resources = getUserResourceList(user);
+
+            const permissions = resources.map(
+                (resource) => resource.permission
+            );
+            const displayPermission = getDisplayPermission(
+                permissions,
+                resourceTotal
+            );
+            return {
+                ...acc,
+                [userId]: {
+                    ...user,
+                    displayPermission,
+                    originalPermission: displayPermission,
+                },
+            };
+        },
+        { ...userMap }
+    );
+};
+
+/**
+ * This will create a mapping where each key is a user ID.  The value is
+ * the user's information from the user-info service with 3 things added:
+ * - originalPermission: contains the original overall permission for the user
+ *
+ * - displayPermission: contains the current overall permission for the user.
+ * If the user's permission gets updated in the sharing dialog, this value will
+ * update.
+ *
+ * - paths, apps, analyses, tools keys for storing the array of resources the
+ * user has or will have permissions on
+ *
+ * @param responses - permission-lister responses
+ * @param userInfos
+ * @param resourceTotal
+ * @returns {{}}
+ */
+export const getUserMap = (responses, userInfos, resourceTotal) => {
+    const userMap = responses.reduce((acc, resp) => {
+        const type = getResponseType(resp);
+        const { permListFn, userIdFn, fmtSharingFn } = getSharingFns(type);
+        return resp[type].reduce(
+            (shares, sharing) => {
+                const permissions = permListFn(sharing);
+
+                return permissions.reduce(
+                    (userMap, permission) => {
+                        const userId = userIdFn(permission);
+                        const currentUser = {
+                            ...(userMap[userId] || userInfos[userId]),
+                        };
```

Comment:
No need to copy here since `currentUser` will be copied below.
```suggestion
                        const currentUser = (userMap[userId] || userInfos[userId]);
```